include/mxnet/ndarray.h

```diff
@@ -309,6 +309,62 @@ class NDArray {
     return true;
   }
   /*!
+   * \brief check whether the matrix format is valid
+   * \param full_check if `True`, rigorous check, O(N) operations
+   *    Otherwise basic check, O(1) operations
+   */
+  inline void check_format(const bool full_check = true) const {
+    auto stype = storage_type();
+    if (stype == kCSRStorage) {
+      CHECK_EQ(shape().ndim(), 2)
+               << "CSR only for two dimension";
+      CHECK(aux_shape(csr::kIdx).ndim() == 1 &&
+            aux_shape(csr::kIndPtr).ndim() == 1 &&
+            storage_shape().ndim() == 1)
+            << "data, indices, and indptr should be 1-D";
+      CHECK_EQ(aux_shape(csr::kIndPtr)[0], shape()[0] + 1)
+               << "index pointer size" << aux_shape(csr::kIndPtr)[0]
+               << "should be " << (shape()[0] + 1);
+      CHECK_EQ(aux_shape(csr::kIdx)[0], storage_shape()[0])
+               << "indices and data should have the same size";
+      auto type = aux_type(csr::kIndPtr);
+      MSHADOW_TYPE_SWITCH(type, IType, {
+        auto dptr = aux_data(csr::kIndPtr).dptr<IType>();
+        CHECK_EQ(dptr[0], 0)
+                 << "index pointer should start with 0";
+        const nnvm::dim_t end = aux_shape(csr::kIndPtr)[0];
+        CHECK_GE(dptr[end-1], aux_shape(csr::kIdx)[0])
```

Comment:
Maybe we should move these to src/operator. `dptr[end-1]` can only be accessed on for cpu ctx. For gpu you have to use kernel launch

Reply:
Got it. Thx.

Reply:
Also this needs to be pushed to engine. Can't do it like this